pennylane_lightning/src/simulator/TransitionKernels.hpp

```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+#include <algorithm>
+#include <complex>
+#include <cstdio>
+#include <random>
+#include <stack>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+#include "StateVectorManagedCPU.hpp"
+#include "StateVectorRawCPU.hpp"
+
+namespace Pennylane {
+
+enum class TransitionKernelType { Local, NonRandom };
+
+template <typename fp_t> class TransitionKernel {
+  public:
+    virtual size_t init_state() = 0;
+    // outputs the next state and the qratio
+    virtual std::pair<size_t, fp_t> operator()(size_t) = 0;
+};
+
+template <typename fp_t>
+class LocalTransitionKernel : public TransitionKernel<fp_t> {
+  private:
+    std::random_device rd_;
+    std::mt19937 gen_;
+    std::uniform_int_distribution<size_t> distrib_num_qubits_;
+    std::uniform_int_distribution<size_t> distrib_binary_;
+    size_t num_qubits_;
+
+  public:
+    LocalTransitionKernel(size_t num_qubits) {
+        num_qubits_ = num_qubits;
+        gen_ = std::mt19937(rd_());
+        distrib_num_qubits_ =
+            std::uniform_int_distribution<size_t>(0, num_qubits - 1);
+        distrib_binary_ = std::uniform_int_distribution<size_t>(0, 1);
+    }
+
+    size_t init_state() { return 0; }
+
+    std::pair<size_t, fp_t> operator()(size_t s1) {
```

Comment:
Again there are some coverage complaints here. Not sure if the c++ tests need to be more explicit, or we need to add some additional information to ensure the coverage is correctly reported.